### PR TITLE
feat(tracing): add ability to sample based on resource

### PR DESCRIPTION
**Pull request by emmettbutler**
This change implements simple resource-based trace sampling by adding the `resource` keyword parameter to `SamplingRule.__init__`. It does not change when `SamplingRule`s are evaluated. This change also removes a redundant test case.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/sampling_rule.py [hunk 1]
```diff
@@ -129,9 +133,13 @@ def matches(self, span):
         :returns: Whether this span matches or not
         :rtype: :obj:`bool`
         """
-        # Our LFU cache expects a single key, convert the
-        # provided Span into a hashable tuple for the cache
-        return self._matches((span.service, span.name))
+        # if we're just interested in root spans and this isn't one, then return False
+        if self.target_span == "root":
+            if span.parent_id is not None:
```
Comment (ZStriker19) [inline ddtrace/sampling_rule.py:138]: "we propose to implement an option in tracers to ignore upstream sampling decision." 

We can likely implement this feature through this part of the code right, by adding a check for whatever config option we receive, or if it's a per rule option we can check the rule.

File: ddtrace/sampling_rule.py [hunk 2]
```diff
@@ -171,4 +179,9 @@ def __eq__(self, other):
         if not isinstance(other, SamplingRule):
             raise TypeError("Cannot compare SamplingRule to {}".format(type(other)))
 
-        return self.sample_rate == other.sample_rate and self.service == other.service and self.name == other.name
+        return (
```
Comment (ZStriker19) [inline ddtrace/sampling_rule.py:182]: Where are we comparing samplingrules?

**Review by ZStriker19 — COMMENTED** [review]
The simplicity of making a simple change.

File: ddtrace/sampling_rule.py (hunk 2 above)
Comment (emmettbutler) [reply ddtrace/sampling_rule.py:182]: I don't know of any code that uses this, but it's part of the public API.

File: ddtrace/sampling_rule.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/sampling_rule.py:138]: > "we propose to implement an option in tracers to ignore upstream sampling decision."

for this, we'd need to ignore it in context propagation extraction? if we ignore it there and then can cause normal root span creation to trigger sampling decision.

I am not confident this is the right call, but what came to mind for me.

Comment (ZStriker19) [reply ddtrace/sampling_rule.py:138]: Ah that makes more sense. Forgot where I was in the code!

**Review by brettlangdon — COMMENTED** [review]
We have some tests for the `DD_SPAN_SAMPLING_RULES` env variable parsing, do we need to add support there for resource?

**Comment by emmettbutler** [comment]
@brettlangdon do you mean `DD_TRACE_SAMPLING_RULES`? This change doesn't touch span sampling.

**Comment by brettlangdon** [comment]
> @brettlangdon do you mean `DD_TRACE_SAMPLING_RULES`? This change doesn't touch span sampling.

Yes, I mistyped. Thank you, E_TOO_MANY_SAMPLING_RULES

File: benchmarks/sampling_rule_matches/scenario.py [hunk 3]
```diff
@@ -27,20 +27,26 @@ class SamplingRules(bm.Scenario):
     num_iterations = bm.var(type=int)
     num_services = bm.var(type=int)
     num_operations = bm.var(type=int)
+    num_resources = bm.var(type=int)
```
Comment (brettlangdon) [inline benchmarks/sampling_rule_matches/scenario.py:30]: you'll need to update the config too I think?

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**